vispy/visuals/volume.py

```diff
@@ -382,11 +430,11 @@
         // Refine search for max value, but only if anything was found
         if ( maxi > -1 ) {
             // Calculate starting location of ray for sampling
+            vec3 loc = start_loc + step * maxi;
             vec3 start_loc_refine = start_loc + step * (float(maxi) - 0.5);
-            loc = start_loc_refine;
 
             // Variables to keep track of current value and where max was encountered
-            vec3 max_loc_tex = start_loc_refine;
+            vec3 max_loc_tex = loc;
```

Comment:
I think this is an edge case bug in the existing shader and it was causing a test failure when I re-sampled the value below (which, again, I'd like to avoid).

The problem is `start_loc_refine` is not the location where the `maxval` was found in `in_loop`, it's half a step back. Then in the refinement loop, this value is only updated if `val > maxval`. The strict comparison (`>`, not `>=`) means this is possibly never hit, and `max_loc_tex` is never updated. In the end this means `max_loc_tex` (used to set the frag depth) is half a step back from the actual location of `maxval`.

The change here is one way to resolve this. Perhaps the better way is re-initialize `maxval` here.

Reply:
I think you're right. If you see how `frag_depth_point` is initialized in `_RYCASTING_SETUP_VOLUME`, you see it's using `front`, which has that 0.5 offset included. Unless it's wrong there? But I think your point makes sense. Is there a way to check this with some test like I did originally in https://github.com/vispy/vispy/pull/2239?

Reply:
I will try - I first need to understand what you did in #2239, haha.

This will only affect the depth buffer, not the color buffer, right? So I think it needs a test with two visuals. Maybe a volume and an image visual where the volume should end up in front, but does not. 🤔

Reply:
> I will try - I first need to understand what you did in https://github.com/vispy/vispy/pull/2239, haha.

I don't remember much, sorry xD

> This will only affect the depth buffer, not the color buffer, right?

Yep, exactly. I'm assuming they will have to be within 0.5 texels from one another for the error to show up?

Reply:
Here's the test I came up with, but I can't quite convince myself it's correct.

```python
@requires_pyopengl()
@requires_application()
def test_mip_depth():
    N = 101
    center = ((N - 1) / 2,) * 3
    sphere_volume = np.meshgrid(
        *[np.linspace(-1, 1, N)] * 3,
    )
    sphere_volume = (np.sqrt(3) - np.linalg.norm(sphere_volume, axis=0)) / np.sqrt(3)
    sphere_volume = sphere_volume.astype(np.float32)

    with TestingCanvas(size=(N, N)) as canvas:
        view = canvas.central_widget.add_view()
        view.camera = "arcball"
        view.camera.fov = 0
        view.camera.depth_value = 1.5 * N
        view.camera.center = center
        view.camera.scale_factor = N

        # render a mip of a soft sphere
        # the maximum of every voxel should be at the center
        scene.visuals.Volume(
            sphere_volume,
            clim=(0, 1),
            cmap="reds",
            method='attenuated_mip',
            attenuation=0.0,
            interpolation='nearest',
            relative_step_size=0.01,
            parent=view.scene,
        )

        # add a solid green plane at the center
        plane = scene.visuals.Plane(
            width=N + 10,
            height=N + 10,
            color=(0, 1.0, 0, 1.0),
            direction="+y",
            parent=view.scene,
        )
        plane.set_transform(
            "st",
            translate=center,
        )

        # no pixels should be all green (showing the plane)
        assert not np.any(
            np.all(canvas.render() == [0, 255, 0, 255], axis=-1)
        )
```

With the current code I get some green pixels and the test fails. I can make the test pass if I change this line:
https://github.com/vispy/vispy/blob/f858c92ab13fb572b2ec070e436c08ea5b1f0944/vispy/visuals/volume.py#L483
to
```
             frag_depth_point = max_loc_tex * u_shape - 0.5;
```

I think this change is what the end of [this comment](https://github.com/vispy/vispy/pull/2239#issuecomment-949960620) was suggesting, but seems to have not been included. 

Which is correct, though, and how can I know? Perhaps we should _expect_ the z-fighting with the plane at the center.

I also was hoping to see the render 100% green if I set the plane forward even a bit, but to make it all green I need to use -10 or so.

Reply:
Something's off, yeah. I made an interactive version of your test, and it's weird to me. Not only that, but I think I'm getting some issues due to the bounding sphere trick you used. Basically, I took the code above and made small changes to make it interactive and more obvious (5 pixels istead of 100):

```py
import numpy as np
from vispy.scene import SceneCanvas, Volume, Plane
N = 5
center = ((N - 1) / 2,) * 3
sphere_volume = np.meshgrid(
    *[np.linspace(-1, 1, N)] * 3,
)
sphere_volume = (np.sqrt(3) - np.linalg.norm(sphere_volume, axis=0)) / np.sqrt(3)
sphere_volume = sphere_volume.astype(np.float32)

canvas = SceneCanvas(size=(N*100, N*100), show=True)
view = canvas.central_widget.add_view()
view.camera = "arcball"
view.camera.fov = 0
view.camera.depth_value = 1.5 * N
view.camera.center = center
view.camera.scale_factor = N

# render a mip of a soft sphere
# the maximum of every voxel should be at the center
Volume(
    sphere_volume,
    clim=(0, 1),
    cmap="reds",
    method='attenuated_mip',
    attenuation=0.0,
    interpolation='nearest',
    relative_step_size=0.01,
    parent=view.scene,
)

# add a solid green plane at the center
plane = Plane(
    width=N + 1,
    height=N + 1,
    color=(0, 1.0, 0, 1.0),
    direction="+y",
    parent=view.scene,
)
plane.set_transform(
    "st",
    translate=center,
)
```

My idea was that this way you should see some z-fighting (as you said, expected). But when you move the plane even slightly, the fighting should be gone. Instead I see some weird glimpes of the plane between the voxels, which I can't quite explain.

Also, you see that round "cut" at the edges of the cube? Suspicious!
![image](https://github.com/user-attachments/assets/c4a5c9e1-3ce6-4cf1-9bd7-b321fb46cd11)

Reply:
That is a bit suspicious, and to me very surprising!

I'm leaving today for vacation but will try to pick this up again next week.

Reply:
No rush! Enjoy your holidays :)

Reply:
I'm tempted to push this `frag_depth_point` into another issue/PR as in the end it's related but not caused by this work.

Reply:
Yep, good idea. Sorry, sometimes I get stuck in "solve everything" mode :P The sphere fix loos great now!